Auto fix outdated links in comments and documentation

WinterSilence opened the pull request:
Adds `\yii\build\helpers\LinkChecker` to checking links (caches results)
Adds `\yii\build\controller\FixLinksController` to finding links, replacing updated links and logging removed links
Adds Github workflow `cron-outdated-links.yml` runing `FixLinksController` every day

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  |  #19314

WinterSilence committed:
Create LinkChecker.php

WinterSilence committed:
Create FixLinksController.php

WinterSilence committed:
Create cron-outdated-links.yml

WinterSilence committed:
Update cron-outdated-links.yml

WinterSilence committed:
Update FixLinksController.php

samdark reviewed (approved):
Looks 👍

samdark commented on build/controllers/FixLinksController.php:
```diff
@@ -0,0 +1,167 @@
+
+/**
+ * Replaces outdated links(3xx response codes) add collect removed links (4xx response codes) in file
+ * `./build/runtime/outdated-links.log`.
```
I don't think collecting links in a file is useful as GitHub action. It should be stdout in order to be visible in results at GitHub.

WinterSilence replied:
@samdark no problem, I was logging removed links because this workflow running by Cron

WinterSilence committed:
Update FixLinksController.php

WinterSilence committed:
Refactor FixLinksController

WinterSilence committed:
Merge branch 'master' into patch-12

WinterSilence committed:
Merge branch 'master' into patch-12

samdark commented on build/controllers/FixLinksController.php:
```diff
@@ -0,0 +1,169 @@
+        }
+        $outdated = ['updated' => [], 'removed' => []];
+        foreach (\array_unique($urls) as $url) {
+            $url = \rtrim($url, ' ],\'`/');
```
`$url` is an array here.

samdark replied:
```
❯ ./build/build fix-links
Exception 'TypeError' with message 'rtrim(): Argument #1 ($string) must be of type string, array given'

in /Users/samdark/src/yii2/build/controllers/FixLinksController.php:158

Stack trace:
#0 /Users/samdark/src/yii2/build/controllers/FixLinksController.php(158): rtrim(Array, ' ],'`/')
#1 /Users/samdark/src/yii2/build/controllers/FixLinksController.php(77): yii\build\controllers\FixLinksController->findOutdatedLinks('/**\n * @link ht...')
#2 /Users/samdark/src/yii2/build/controllers/FixLinksController.php(41): yii\build\controllers\FixLinksController->fixSources()
#3 [internal function]: yii\build\controllers\FixLinksController->actionIndex()
#4 /Users/samdark/src/yii2/framework/base/InlineAction.php(57): call_user_func_array(Array, Array)
#5 /Users/samdark/src/yii2/framework/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#6 /Users/samdark/src/yii2/framework/console/Controller.php(182): yii\base\Controller->runAction('', Array)
#7 /Users/samdark/src/yii2/framework/base/Module.php(552): yii\console\Controller->runAction('', Array)
#8 /Users/samdark/src/yii2/framework/console/Application.php(180): yii\base\Module->runAction('fix-links', Array)
#9 /Users/samdark/src/yii2/framework/console/Application.php(147): yii\console\Application->runAction('fix-links', Array)
#10 /Users/samdark/src/yii2/framework/base/Application.php(384): yii\console\Application->handleRequest(Object(yii\console\Request))
#11 /Users/samdark/src/yii2/build/build(40): yii\base\Application->run()
#12 {main}
```

WinterSilence replied:
@samdark fixed

WinterSilence committed:
Fix FixLinksController::findOutdatedLinks()

WinterSilence committed:
Merge branch 'master' into patch-12

samdark commented:
Still errors:

```
❯ ./build/build fix-links
PHP Warning 'yii\base\ErrorException' with message 'Undefined array key "http://www.yiiframework.com"'

in /Users/samdark/src/yii2/build/helpers/LinkChecker.php:72

Stack trace:
#0 /Users/samdark/src/yii2/build/helpers/LinkChecker.php(72): yii\base\ErrorHandler->handleError(2, 'Undefined array...', '/Users/samdark/...', 72)
#1 /Users/samdark/src/yii2/build/helpers/LinkChecker.php(37): yii\build\helpers\LinkChecker->checkUrl('http://www.yiif...')
#2 /Users/samdark/src/yii2/build/controllers/FixLinksController.php(158): yii\build\helpers\LinkChecker->check('http://www.yiif...')
#3 /Users/samdark/src/yii2/build/controllers/FixLinksController.php(77): yii\build\controllers\FixLinksController->findOutdatedLinks('/**\n * @link ht...')
#4 /Users/samdark/src/yii2/build/controllers/FixLinksController.php(41): yii\build\controllers\FixLinksController->fixSources()
#5 [internal function]: yii\build\controllers\FixLinksController->actionIndex()
#6 /Users/samdark/src/yii2/framework/base/InlineAction.php(57): call_user_func_array(Array, Array)
#7 /Users/samdark/src/yii2/framework/base/Controller.php(178): yii\base\InlineAction->runWithParams(Array)
#8 /Users/samdark/src/yii2/framework/console/Controller.php(182): yii\base\Controller->runAction('', Array)
#9 /Users/samdark/src/yii2/framework/base/Module.php(552): yii\console\Controller->runAction('', Array)
#10 /Users/samdark/src/yii2/framework/console/Application.php(180): yii\base\Module->runAction('fix-links', Array)
#11 /Users/samdark/src/yii2/framework/console/Application.php(147): yii\console\Application->runAction('fix-links', Array)
#12 /Users/samdark/src/yii2/framework/base/Application.php(384): yii\console\Application->handleRequest(Object(yii\console\Request))
#13 /Users/samdark/src/yii2/build/build(40): yii\base\Application->run()
#14 {main}
```

WinterSilence committed:
Merge branch 'master' into patch-12

WinterSilence committed:
Update LinkChecker.php

WinterSilence committed:
Fix FixLinksController

WinterSilence commented:
@samdark нет бы, помочь.. поправил

WinterSilence committed:
Update FixLinksController.php

WinterSilence committed:
Filter test/dummy links in `LinkChecker::check()`

samdark commented:
It takes forever to check all links but it seems it works. There are some issues though:

1. https://www.yiiframework.com/extension/yiisoft/yii2-twig/doc/guide/ is replaced with https://www.yiiframework.com/extension/yiisoft/yii2-twig/doc/guide/2.4/en/. First link always leads to latest version.
2. Same for https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html.
3. https://developer.mozilla.org/en-US/docs/HTTP/Access_control_CORS was replaced with /en-us/docs/web/http/cors.
4. https://www.microsoft.com/sql-server/ was replaced with https://www.microsoft.com/en-us/sql-server?rtc=1/ i.e. it's language specific URL now instead of generic routing page.
5. https://www.oracle.com/database/ was replaced with /database/.

samdark commented:
Judging on these, I'm not sure it's a good idea to add it to pipeline.

WinterSilence commented:
> https://www.yiiframework.com/extension/yiisoft/yii2-twig/doc/guide/ is replaced with https://www.yiiframework.com/extension/yiisoft/yii2-twig/doc/guide/2.4/en/. First link always leads to latest version.

Permanent redirects codes: 301 and 308.
Temporary redirects: 302 and 307.
i.е. `https://www.yiiframework.com` return invalid code. We can skip temporary redirects.

> https://www.microsoft.com/sql-server/ was replaced with https://www.microsoft.com/en-us/sql-server?rtc=1/ i.e. it's language specific URL now instead of generic routing page.

English manual must contains links to `en` pages, Russian manual must contains links to `ru` pages and etc.

WinterSilence committed:
Merge branch 'master' into patch-12

WinterSilence commented:
@samdar and why `https://www.yiiframework.com` redirect twice?

WinterSilence committed:
Add LinkChecker::$skipHttpCodes

Ignores some redirects

WinterSilence commented:
@samdark partially fixed
> I'm not sure it's a good idea to add it to pipeline to automatically fix things.

at first, need collect links like as `https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html` and append to ignore list, at next callings we replaces permanently redirected links .

WinterSilence committed:
Merge branch 'master' into patch-12

WinterSilence committed:
Update FixLinksController

Add option `FixLinksController::$safeMode`
Fix trimming in `findOutdatedLinks()`

WinterSilence committed:
Update cron-outdated-links.yml

WinterSilence commented:
@samdark add safe mode to collect links without changes in files

WinterSilence committed:
Merge branch 'master' into patch-12

WinterSilence committed:
Merge branch 'master' into patch-12

WinterSilence committed:
Merge branch 'master' into patch-12

WinterSilence committed:
Update FixLinksController.php

WinterSilence committed:
Update cron-outdated-links.yml

WinterSilence committed:
Merge branch 'master' into patch-12
